fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java

```diff
@@ -0,0 +1,523 @@
+ * eg.
+ * t1.id = t2.id and t2.id = t3.id and t3.id = 100;
+ * -->
+ * t1.id = 100 and t2.id = 100 and t3.id = 100;
+ *
+ * 1. Register a new rule InferFiltersRule and add it to GlobalState.
+ * 2. Traverse Conjunct to construct on/where equivalence connection, numerical connection and isNullPredicate.
+ * 3. Use Warshall to infer all equivalence connections
+ * 4. Construct additional numerical connections and isNullPredicate
+ */
+public class InferFiltersRule implements ExprRewriteRule {
+    private final static Logger LOG = LogManager.getLogger(InferFiltersRule.class);
+    public static InferFiltersRule INSTANCE = new InferFiltersRule();
+
+    @Override
+    public Expr apply(Expr expr, Analyzer analyzer, ExprRewriter.ClauseType clauseType) throws AnalysisException {
+        if (expr == null) {
+            return null;
+        }
+        
+        if (!analyzer.enableInferPredicate() || clauseType == ExprRewriter.ClauseType.OTHERCLAUSE) {
+            return expr;
+        } 
+
+        // slotEqSlotExpr: Record existing and infer equivalent connections
+        List<Expr> slotEqSlotExpr =
+                (clauseType == ExprRewriter.ClauseType.ONCLAUSE) ? analyzer.getOnSlotEqSlotExpr() : new ArrayList<>();
+
+        // slotEqSlotDeDuplication: De-Duplication for slotEqSlotExpr
+        Set<Pair<Expr, Expr>> slotEqSlotDeDuplication =
+                (clauseType == ExprRewriter.ClauseType.ONCLAUSE) ? analyzer.getOnSlotEqSlotDeDuplication() : Sets.newHashSet();
+
+        // slotToLiteralExpr: Record existing and infer expr which slot and literal are equal
+        List<Expr> slotToLiteralExpr =
+                (clauseType == ExprRewriter.ClauseType.ONCLAUSE) ? analyzer.getOnSlotToLiteralExpr() : new ArrayList<>();
+
+        // slotToLiteralDeDuplication: De-Duplication for slotToLiteralExpr
+        Set<Pair<Expr, Expr>> slotToLiteralDeDuplication =
+                (clauseType == ExprRewriter.ClauseType.ONCLAUSE) ? analyzer.getOnSlotToLiteralDeDuplication() : Sets.newHashSet();
+
+        // newExprWithState: just record infer expr which slot and literal are equal and which is not null predicate
+        // false : Unexecutable intermediate results will be produced during the derivation process.
+        // true : The new expr will be add to expr.
+        List<Pair<Expr, Boolean>> newExprWithState = new ArrayList<>();
+
+        // isNullExpr: Record existing and infer not null predicate
+        List<Expr> isNullExpr =
+                (clauseType == ExprRewriter.ClauseType.ONCLAUSE) ? analyzer.getOnIsNullExpr() : new ArrayList<>();
+
+        //isNullDeDuplication: De-Duplication for isNullExpr
+        Set<Expr> isNullDeDuplication =
+                (clauseType == ExprRewriter.ClauseType.ONCLAUSE) ? analyzer.getOnIsNullDeDuplication() : Sets.newHashSet();
+
+
+        // exprToInteger/integerToExpr: function is easy to build warshall and newExprWithState
+        Map<Expr, Integer> exprToInteger = new HashMap<>();
```

Comment:
What's the meaning of integer in `exprToInteger` and `integerToExpr`? Names of variables should be as specific as possible.

Reply:
already edited